### ISSUE TITLE
Error: "The entity type 'DataProtectionKey' requires a primary key" when using .Net 7 previews, EF Core and Linux

### ISSUE THREAD

**Issue by Webreaper** [dotnet/efcore#28444]

**Comment by Webreaper** [comment]
Thanks for moving. I originally went to open this in EFCore, but then decided it was more of a aspnetcore issue. 🤣

Just to sanity check, this morning I ran `dotnet ef migrations add` to see if there was some fundamental change within the `DataProtectionKeys` structure that might be teased out into a migration to fix this issue, but (as probably expected) it didn't have any effect.

**Comment by Webreaper** [comment]
Okay, so additional info. I've managed to workaround this by explicitly adding the key with my `OnModelCreating` method:

```
   protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            // Potential fix for https://github.com/dotnet/efcore/issues/28444
            var dpk = modelBuilder.Entity<DataProtectionKey>();
            dpk.HasKey(x => x.Id);
```
With this, the app starts up correctly, as the key has been correctly added to the entity. 

So I guess the questions are:

1. Should developers need to explicitly declare the key in their `OnModelCreating` code? if so, I've never seen that in the documentation for that function (e.g., it's not mentioned in the [Configure ASP.NET Core Data Protection](https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/overview?view=aspnetcore-6.0) docs.
2. Why does it work on OSX? Is there some introspection that's failing on Linux (or in Docker) that's affecting the entity key declaration? I wondered if something was being stripped out as part of the single-file publishing/trimming process, but I've tried turning off `/p:PublishSingleFile=true /p:PublishTrimmed=true` just in case, and it didn't seem to make a difference.

Let me know if you need any more details.

**Comment by ajcvickers** [comment]
@Webreaper Sounds like this could be related to trimming. I suspect when the `Id` property is not referenced directly, then on platforms that turn on trimming by default it is getting trimmed out.

**Comment by Webreaper** [comment]
Yes, it could be feasible. I do deploy as a single-file executable, and usually have trimming etc enabled:
```
# dotnet publish Damselfly.Web -r $runtime -f net${dotnetversion} -c Release --self-contained true /p:Version=$version /p:PublishSingleFile=true /p:PublishTrimmed=true /p:IncludeNativeLibrariesForSelfExtract=true
```

but I disabled trimming to see if it would have an effect (or would fix this problem). 

Also, I had trimming disabled for Preview 5 due to a related bug (https://github.com/dotnet/runtime/issues/70758) but disabling trimming doesn't prevent this issue from happening.

**Comment by ajcvickers** [comment]
@Webreaper Are you sure `Microsoft.AspNetCore.DataProtection.EntityFrameworkCore` is not being trimmed?

**Comment by Webreaper** [comment]
If trimming is not enabled in the build/publish, how would it be trimmed? Or is there some magic I'm missing?

**Comment by Webreaper** [comment]
Maybe @richlander might have an idea?

**Comment by Webreaper** [comment]
Also, if this was a trimming error, it wouldn't explain why it doesn't happen before EFCore 7 preview 5.

**Comment by ajcvickers** [comment]
@Webreaper What happens if you remove this line:

```C#
 dpk.HasKey(x => x.Id);
```

And replace it with:

```C#
Console.WriteLine(new DataProtectionKey().Id);
```

Or anything else that references the `Id` property without configuring it in EF model? Does the code still work like it does when using ` dpk.HasKey(x => x.Id);`? If so, then that implies that just referencing the `Id` property is enough to have EF find it. If it doesn't then, we're going to have to start reducing the complexity in the repro until we can figure out the root cause.

**Comment by Webreaper** [comment]
Sure. Will give it a try.

**Comment by richlander** [comment]
Can you share the `dotnet build` command you are using and any relevant properties ion your project file?

**Comment by Webreaper** [comment]
Here's the `dotnet publish` command:

https://github.com/Webreaper/Damselfly/blob/develop/scripts/makeserver.sh#L33

It evaluates to:
```
dotnet publish Damselfly.Web -r linux-x64 -f net7.0 -c Release --self-contained true /p:Version=3.1.0 /p:EnableCompressionInSingleFile=true /p:PublishSingleFile=true /p:PublishTrimmed=true /p:IncludeNativeLibrariesForSelfExtract=true
```

I also have [Directory.Build.props](https://github.com/Webreaper/Damselfly/blob/develop/Directory.Build.props), and [here's the project file](https://github.com/Webreaper/Damselfly/blob/develop/Damselfly.Web/Damselfly.Web.csproj) which doesn't have anything fancy in terms of properties. Let me know if I can point you at anything else.

Note that [this commit](https://github.com/Webreaper/Damselfly/commit/06c86c218773a3a0c9a3592bcc5a6c88ddf8c41f) seemed to fix the problem per earlier comments. So it's not a huge issue, but it's somewhat confusing as I don't know why it should be needed (I would presume that `DataProtectionKey.Id` is already explicitly marked with `[Key]`).

**Comment by richlander** [comment]
Do you get the error just the following command?

```bash
dotnet publish Damselfly.Web -r linux-x64 -f net7.0 -c Release
```

**Comment by Webreaper** [comment]
Will try. Can't remember if my deployment scripts will break if it's not a single file deployment, so I might have to tweak.

**Comment by Webreaper** [comment]
Okay, so that's confirmed. This breaks:
'''
dotnet publish Damselfly.Web -r linux-x64 -f net7.0 -c Release --self-contained true /p:Version=3.1.0 
/p:PublishSingleFile=true /p:IncludeNativeLibrariesForSelfExtract=true
'''
This works:
```
dotnet publish Damselfly.Web -r linux-x64 -f net7.0 -c Release
```
So it is to do with single-file deployment and/or trimming. What combination do you want me to try next?

**Comment by Webreaper** [comment]
Okay, so adding `self-contained true` works. 

Then adding '/p:PublishSingleFile=true', and it's also fine.

Adding `/p:PublishTrimmed=true` causes it to break. So it looks like it is related to trimming.

**Comment by richlander** [comment]
Perfect. That's what we wanted to find out. Let's see if we can get some help.

@agocke @vitek-karas

**Comment by ajcvickers** [comment]
I think the behavior here is correct. The application is mapping a type (`DataProtectionKey`) from one of our libraries (`Microsoft.AspNetCore.DataProtection`) as an entity type in the application's model. This is a very unusual thing to do, but in this case is not totally unreasonable because `DataProtectionKey` could conceptually be considered an entity type.

From a trimming perspective, we don't have any knowledge that the `Id` property is going to be mapped by convention by EF Core, and hence trimming it out is reasonable since it is never explicitly used anywhere in the code. This can be fixed by explicitly mapping that the `Id` property as a key, which causes it to be referenced and not trimmed.

**Comment by Webreaper** [comment]
Makes sense. I guess the thing that confuses me is that this issue doesn't happen at with .Net 6, and (I'm pretty certain, but can confirm if you need to) that it didn't happen with the .Net 7 before preview 5. So that implies a change in behaviour that might confuse people when 7 GA is released.

Perhaps as you guys are turning the screws on the trimming efficiency it's resulted in the change? But possibly you might be able to put some hints to stop it getting trimmed? After all, I'm not doing anything explicit with the Id field, I'm just using the standard EFCore Data Protection Keys functions - which made this error super confusing from my perspective at first because it's not even referenced within my app - it's something internal to Asp.Net identity.

**Comment by roji** [comment]
> I guess the thing that confuses me is that this issue doesn't happen at with .Net 6, and (I'm pretty certain, but can confirm if you need to) that it didn't happen with the .Net 7 before preview 5. So that implies a change in behaviour that might confuse people when 7 GA is released.

That is likely because of [this change in .NET 7](https://docs.microsoft.com/en-us/dotnet/core/compatibility/deployment/7.0/trim-all-assemblies).

**Comment by richlander** [comment]
> this can be fixed by explicitly mapping that the Id property as a key

That was directed at you @Webreaper.

**Comment by Webreaper** [comment]
Got it.

I'm fine with this and the fix works well. I'm just wondering if there's any hints or docs you could put together because other people will get caught by this; while the fix to add the reference is straightforward, it might catch people unawares to have to add a `HasKey` call for a model they may not even be aware they're using. Although hopefully googling this issue will help, or I can write an SO question and answer for it.... 😁

**Comment by Webreaper** [comment]
Oh, and I think the "on Linux" part may be misleading; I may not see it on Mac because that's my Dev env so I don't usually publish as a single file there. I'll test it when I get 5 mins, and if it's consistent I'll change the issue title to "when publishing trimmed" so finding the issue and solution is easier for anyone who is hit by it.

**Comment by roji** [comment]
We do plan to make the experience better when trimming. For example, we could annotate EF APIs to preserve all public properties on entity CLR types. We're not sure we'll be able to do significant trimming work for 7.0, but at least some guidance/documentation does make sense.

**Comment by Webreaper** [comment]
Additional info on the cause of this here, for anyone landing on this issue: https://devblogs.microsoft.com/dotnet/announcing-dotnet-7-preview-7/#trimming-and-nativeaot-all-assemblies-trimmed-by-default
